src/wallet/rpcwallet.cpp

```diff
@@ -2047,25 +2061,21 @@ UniValue ListReceived(const UniValue& params, bool by_label)
         }
     }
 
-    // Reply
-    UniValue ret(UniValue::VARR);
-    std::map<std::string, tallyitem> label_tally;
     // Create mapAddressBook iterator
     // If we aren't filtering, go from begin() to end()
-    auto start = pwalletMain->mapAddressBook.begin();
-    auto end = pwalletMain->mapAddressBook.end();
+    auto itAddrBook = pwalletMain->NewAddressBookIterator();
     // If we are filtering, find() the applicable entry
     if (has_filtered_address) {
-        start = pwalletMain->mapAddressBook.find(filtered_address);
-        if (start != end) {
-            end = std::next(start);
-        }
+        itAddrBook.SetFilter(filtered_address);
     }
 
-    for (auto item_it = start; item_it != end; ++item_it)
-    {
-        const CTxDestination& address = item_it->first;
-        const std::string& label = item_it->second.name;
+    // Reply
+    UniValue ret(UniValue::VARR);
+    std::map<std::string, tallyitem> label_tally;
+
+    while(itAddrBook.HasNext()) {
```

Comment:
Same here. There is a `continue` below (line 2081) that make this an infinite loop.
This is actually hit during `wallet_listreceivedby.py` (you can see travis consistently timing out).